Feature kafka monitoring

ZhuWang1112 opened the pull request:
- [ ] If this is non-trivial feature, paste the links/URLs to the design doc.

- [x] Update the documentation to include this new feature.

- [x] Tests(including UT, IT, E2E) are added to verify the new feature.

- [x] If it's UI related, attach the screenshots below.

- [ ] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #<issue number>.

- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/docs/en/changes/changes.md).

Kafka Root Dashboard
<img width="1917" alt="kafka-root" src="https://github.com/apache/skywalking/assets/97995576/69d5b8ea-2940-4a32-af3d-a361c890c0ec">


Kafka Cluster Dashboard
<img width="1913" alt="kafka-cluster" src="https://github.com/apache/skywalking/assets/97995576/3a8688e6-294c-4418-92ef-2bcce317a77f">

<img width="1171" alt="clusterOverView" src="https://github.com/apache/skywalking/assets/97995576/d1851317-df53-4718-ac14-832fb13db510">

Kafka Broker Dashboard
<img width="1905" alt="kafka-broker0" src="https://github.com/apache/skywalking/assets/97995576/1cb52145-31fc-4b3e-9c6b-e0a2cfb44121">
<img width="1909" alt="kafka-broker1" src="https://github.com/apache/skywalking/assets/97995576/e60bff91-ebc2-4edc-9f7e-f615f9a2625d">

wankai123 commented on test/e2e-v2/cases/kafkamq/docker-compose.yml:
```diff
@@ -0,0 +1,137 @@
+
+version: '3'
+
+services:
```
missing oap service deployment

wankai123 commented on test/e2e-v2/cases/kafkamq/otel-collector-config.yaml:
```diff
@@ -0,0 +1,56 @@
+
+exporters:
+  otlp:
+    # endpoint: "10.0.0.120:11800" # The OAP Server address
+    endpoint: "192.168.0.120:11800"
```
```suggestion
    # The OAP Server address
    endpoint: oap:11800
```

wankai123 commented:
Please don't forget the document.

wu-sheng commented:
1. skywalking-ui submodule seems not to be updated correctly. You should not change the commit ID of this submodule. Please align with the master branch.
2. Your newly added e2e is not in the CI pipeline, so it would not run even if I approved this. Please add them to the proper CI control file.

If you are not sure what I meant, please connect with your mentor. Thanks.

wu-sheng commented:
`Conflicting files` must be resolved, otherwise, CI can't run.

ZhuWang1112 commented:
> `Conflicting files` must be resolved, otherwise, CI can't run.

wu-sheng commented:
Why close?

ZhuWang1112 commented:
> `Conflicting files` must be resolved, otherwise, CI can't run.
Thanks!  I'm still revising some previous content. There are still some problems with e2e's current local testing. I will fix the problem you mentioned in time

ZhuWang1112 commented:
> Why close?
Sorry for the mistake.

wuwen5 commented on oap-server/server-starter/src/main/resources/application.yml:
```diff
@@ -393,7 +393,7 @@ query:
     maxQueryComplexity: ${SW_QUERY_MAX_QUERY_COMPLEXITY:3000}
     # Allow user add, disable and update UI template
-    enableUpdateUITemplate: ${SW_ENABLE_UPDATE_UI_TEMPLATE:false}
+    enableUpdateUITemplate: ${SW_ENABLE_UPDATE_UI_TEMPLATE:true}
```
Why modify the default values? Please confirm if this modification is related to this feature.
If necessary, please modify the corresponding document. (`skywalking/docs/en/setup/backend/Configuration vocabulary.md`)

wu-sheng replied:
This should not open by default.
I believe this is a local change, please revert.

ZhuWang1112 replied:
Sorry, this is just for local testing convenience. I will turn it off before the final version

wu-sheng replied:
The best practice is, you should revert this, and change the local system environment variable to set as true.

wu-sheng commented:
If you are ready, please make this as `ready for review`. I would approve your tests on CI.

wu-sheng commented:
> test/e2e-v2/cases/kafka/kafka-monitoring/jmx_prometheus_javaagent-0.18.0.jar

This can't be accepted. There is no binary jar allowed in ASF source code repository and final release.
If you need to it, please download in the test stage.

ZhuWang1112 commented:
> If you are ready, please make this as `ready for review`. I would approve your tests on CI.

Thank you so much! I fixed the download method of javagent.jar and have opened this PR

wu-sheng commented:
Is the last commit breaking something? Most tests passed, but fail this time.

ZhuWang1112 commented:
> Is the last commit breaking something? Most tests passed, but fail this time.

 I thought I didn't make changes to the other codes. I only added the missing license headers, removed a YML file to download it during build time, closed the enableUpdateUITemplate, and cleaned up some comments. I noticed that the Kafka e2e tests I wrote passed successfully.

wu-sheng commented:
Let me rerun cases.

wankai123 commented on oap-server/server-starter/src/main/resources/otel-rules/kafka/kafka-cluster.yaml:
```diff
@@ -0,0 +1,67 @@
+    exp: kafka_controller_controllerstats_uncleanleaderelections_total.sum(['cluster', 'broker']).rate('PT1M')
+
+  - name: max_lag
+    exp: kafka_server_replicafetchermanager_maxlag.sum(['cluster', 'broker'])
+
+
+
+
+
+
+
+
+
```
```suggestion
    exp: kafka_server_replicafetchermanager_maxlag.sum(['cluster', 'broker'])

```

wankai123 commented on test/e2e-v2/cases/kafka/kafka-monitoring/docker-compose.yml:
```diff
@@ -0,0 +1,131 @@
+    depends_on:
+      zookeeper:
+        condition: service_healthy
+    # healthcheck:
+    #   test: ["CMD", "kafka-topics.sh", "--list", "--zookeeper", "zookeeper:2181"]
+    #   interval: 5s
+    #   timeout: 60s
+    #   retries: 120
```
```suggestion
    health check:
      test: ["CMD", "kafka-topics.sh", "--list", "--zookeeper", "zookeeper:2181"]
      interval: 5s
      timeout: 60s
      retries: 120
```

wankai123 commented on test/e2e-v2/cases/kafka/kafka-monitoring/docker-compose.yml:
```diff
@@ -0,0 +1,131 @@
+      - KAFKA_OPTS=-javaagent:/etc/jmx_prometheus_javaagent-0.18.0.jar=7072:/etc/kafka-2_0_0.yml
+    depends_on:
+      zookeeper:
+        condition: service_healthy
```
```suggestion
        condition: service_healthy
    health check:
      test: ["CMD", "kafka-topics.sh", "--list", "--zookeeper", "zookeeper:2181"]
      interval: 5s
      timeout: 60s
      retries: 120
```

wankai123 commented on oap-server/server-starter/src/main/resources/ui-initialized-templates/kafka/kafka-cluster.json:
```diff
@@ -0,0 +1,232 @@
+					"i": "13",
+					"type": "Widget",
+					"metricTypes": ["readLabeledMetricsValues"],
+					"metrics": ["meter_kafka_partition_count"],
```
we should use MQE for queries in the new feature

wu-sheng commented:
Few tests still fail, I am not sure whether it is relative. Let's check again after you fix your mentor's comments.

ZhuWang1112 commented:
> Few tests still fail, I am not sure whether it is relative. Let's check again after you fix your mentor's comments.

I have reviewed my commits and did not find any changes to someone else's code. It might be a github network or system issue?

wu-sheng commented:
GHA has unstable issues last week. I just approved your latest commit, let's see how it goes this time.

wankai123 commented:
It looks like your UI screenshot is not the latest, please update. And be consistent with the existing UI dashboard style.

wu-sheng commented on .github/workflows/skywalking.yaml:
```diff
@@ -676,7 +676,9 @@ jobs:
           path: docker-images
       - name: Load docker images
-        run: find docker-images -name "*.tar" -exec docker load -i {} \;
+        run: |
+          find docker-images -name "*.tar" -exec docker load -i {} \;
+          find docker-images -name "*.tar" -exec rm {} \;
```
@ZhuWang1112 We helped you to fix this. New downloads seem to be over the volume size limitation, which causes the failure.

ZhuWang1112 replied:
Really appreciate your help, and I will be mindful of the volume size limitation in the future.

wankai123 reviewed (approved):
LGTM

wu-sheng commented:
@ZhuWang1112 Are you going to work on the showcase repo to make this available on the demo? If so, we already have elasticsearch deployment in the showcase I guess @wankai123 Is that possible to monitoring our own elasticsearch?

ZhuWang1112 commented:
> Are you going to work on the showcase repo to make this available on the demo?

Thanks for merging my PR! Yep, I'm on it - will add it to the showcase soon. Appreciate your support!
